src/decode.rs

```diff
@@ -24,8 +25,8 @@ fn decode_len<R: Read>(reader: &mut R) -> Result<usize, Error> {
 }
 
 /// Decode a `Value` from avro format given its `Schema`.
-pub fn decode<R: Read>(schema: &Schema, reader: &mut R) -> Result<Value, Error> {
-    match *schema {
+pub fn decode<R: Read>(schema: &Arc<Schema>, reader: &mut R, context: &mut SchemaParseContext) -> Result<Value, Error> {
```

Comment:
What I have done was to take extend the previous PR trying to accomplish this which made the transition from &Schema to Rc<Schema> when we started temporarily additionally storing the schemata in the SchemaParseContext, and then I changed  Rc to Arc because of the new requirement for Send to be implemented here: 

https://gh/stew/avro-rs/blob/feature%2Frefernce-types/src/schema.rs#L1234

This is an area where my lack of experience with Rust is going to be obvious, so I welcome any advice,  but it does seem to me that we should be able to go back to using references  if we track the lifetimes all they way through so that we  are proving that the additional references we are keeping only last for the lifetime of the decode operation. That was a  battle I wasn't ready to wage with the typechecker when I was trying to get all this proved out, but I can take another stab at it.